File: Data Structures & Algorithms/reorder-linked-list/submission-5.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        if not head:
            return
        # find middle
        length = 0
        curr = head
        while curr is not None:
            curr = curr.next
            length += 1
        if length == 1:
            return
        # find middle node
        # infinite loop to check
        pos = 0
        curr = head
        print(length//2)
        while pos < (length // 2) - 1:
            curr = curr.next
            pos += 1
        print(pos, curr.val)
        last = curr
        curr = curr.next
        mid = curr
        last.next = None
        # reverse second half of list
        prev = None
        while curr is not None:
            temp = curr.next
            curr.next = prev
            prev = curr
            curr = temp
        # now merge both lists
        curr_left = head
        curr_right = prev
        while curr_left is not None:
            temp_left = curr_left.next
            temp_right = curr_right.next
            curr_left.next = curr_right
            if temp_left:
                curr_right.next = temp_left
            curr_left = temp_left
            curr_right = temp_right







        

File: Data Structures & Algorithms/merge-two-sorted-linked-lists/submission-5.js
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     constructor(val = 0, next = null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */

class Solution {
    /**
     * @param {ListNode} list1
     * @param {ListNode} list2
     * @return {ListNode}
     */
    mergeTwoLists(list1, list2) {
        let out;
        if(list1 && list2){
            if(list1.val < list2.val){
                out = list1
                list1 = list1.next
            }
            else{
                out = list2
                list2 = list2.next
            }     
        }
        else if(list1){
            out = list1
            list1 = list1.next
        }
        else if(list2){
            out = list2
            list2 = list2.next
        }
        else{
            return null
        }
        let head = out
        while(list1 != null || list2!= null){
            if(list1 && list2){
                if(list1.val < list2.val){
                    out.next = list1
                    list1 = list1.next
                }
                else{
                    out.next = list2
                    list2 = list2.next
                }
            }
            else if(list1){
                out.next = list1
                list1 = list1.next
            }
            else{
                out.next = list2
                list2 = list2.next
            }
            out = out.next
        }
        return head
    }
}
